### ISSUE TITLE
Beta version treats 'not' of a list differently

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap#2594]
In the release

![image](https://user-images.githubusercontent.com/2138696/82908022-f8ba6800-9f99-11ea-8c19-4add234eeec0.png)

reports false. In 6.0.0 beta it reports a list of two falses.

'Find first item ...' reports false or a list. So one is forced to not equal to false instead of just not.

**Comment by jmoenig** [comment]
yes, that's part of an ongoing discussion about APL style "hyperization" of certain blocks. I'm also making it so that Boolean input slots now require a Boolean input (or a list of Boolean values if they're "hyper"-enabled), otherwise they will throw an error. This means you won't be able to use ```not``` for type-casting a non-falsy value into a Boolean anymore.

Regarding ```find``` : I'm very unhappy about the choice to let it report "false" in case it doesn't find an item that satisfies the test. Thinking about alternatives, possibly a third unevaluated input to be reported if none is found.

**Comment by ToonTalk** [comment]
Many languages support the provision of a default value if something isn't found. I like it. But I'm unclear about the "unevaluated" aspect. 

Is there a way for Hyper block support to be enabled in general but not when loading an old project that may depend upon the old behaviour?

**Comment by jmoenig** [comment]
"unevaluated" so that if, for example, you want to throw an error in case nothing is found that error only gets triggered after the search is over, and not always before evaluating the block (as in applicative order of operations).

Hmm... there is now a way that lets you disable hyper-blocks in the settings menu on a per-project basis. Would that help?

**Comment by ToonTalk** [comment]
Setting it per-project would help but for projects saved before this was possible I think the default should be that it is off.

**Comment by jmoenig** [comment]
So, my changes to conditionals - disallowing non-Boolean truthy and falsy values - are another problem for these projects, right?

**Comment by ToonTalk** [comment]
I think projects authored by people used to Lisp/Scheme or JavaScript are likely to rely upon mixing booleans and other values. And I think nearly all programmers think it odd to see` x = False` in code (or  `x = True`).

**Comment by brianharvey** [comment]
I vote no on adding an input to FIND FIRST providing the value to report 
if nothing matches.  This is the sort of thing that you usually hate, 
Jens: hairing up the interface for naive users, for the sake of 
hotshots.  And FIND FIRST isn't the only block that reports failure some 
of the time! ITEM, for example, reports an empty string if the index is 
bigger than the length of the list.  We should think this through 
carefully before changing FIND FIRST.

Reporting False was my idea.  I designed it that way for two reasons: 
(1) That's what Scheme does. (2) An empty list as a list element is more 
common, I think, than a Boolean in a list.  So there's a lower chance of 
ambiguity.  I think probably the right thing is to invent a Failure 
object and use it consistently -- but arithmetic function domain errors 
now correctly report NaN, so maybe what we should do is use NaN as the 
general failure object?

Anyway we need a longer conversation on this topic than I have time for 
this morning.  Please don't do anything rash.

And it sounds as if NOT is another candidate for my SCALAR function block.

I'm starting to move my view in the direction of saying 0, False, and 
the empty list are false; everything else is true.  But maybe that 
category should include the empty string, and/or maybe it shouldn't 
include the empty list.  Zero because of APL Boolean algebra.  Empty 
list for the sake of NOT.  But this is just the early glimmer of a 
position; I'm far from being confident about it.

**Comment by ToonTalk** [comment]
While I have decided what I think about the NaN suggestion, I find this JavaScript quirk amusing:

```
typeof NaN
"number"
```

In my projects I don't need the full generality of `find` so I use either of the following:

![image](https://user-images.githubusercontent.com/2138696/82923816-0ed22380-9fae-11ea-887d-c2c781300202.png)

![image](https://user-images.githubusercontent.com/2138696/82923954-3de89500-9fae-11ea-8806-babde2fec24b.png)

**Comment by jguille2** [comment]
Hi!
Not easy... but I only want to share what I think.
I also don't like "find" reporting "false". For me, it has no sense (semantically).
- We have "contains" block for this, reporting true or false. It's nice.
- But "keep" reports an empty list if there are no matches
- and also "item (4) of (a list of three items)" reports an empty value.

So I like "find" also reporting an empty value if there is no matches.

And then, we can convert easily empty values to false boolean if you want

![emptyToBoolean](https://user-images.githubusercontent.com/2926315/82928804-04e6fb00-9f83-11ea-8389-e9f1e4efe4df.png)

Joan

**Comment by brianharvey** [comment]
You have to be careful about the phrase "empty value."  There is an 
empty /list/, and there is an empty /string/.  I think even in JS those 
aren't the same.  So you can't just say "failure is an empty value"; you 
have to say which, and if it's different in different situations that'll 
just confuse the users, imho.

**Comment by jmoenig** [comment]
hmm... not easy, indeed. But @jadga-h also told me she would greatly prefer ```find``` to report the same empty value (which is the empty string, but can be cast to the number zero) as we're also reporting for out-of-bounds indices.

The issue about `false` that we've been observing is that users are tempted to put the `find` reporter into the test-slot of conditionals, because sometimes they report a Boolean.

So, what I would really like to avoid, but can't without the third slot is having to

1) create an extra variable for the result of ```find```
2) user ```find``` inside an expression that assigns its result to that variable
3) test whether the value of the variable is equals to false (or to the desired data type)

But, to be fair, this wouldn't be easy with a 3-input version of ```find```  either, unless the third input is a ring, which in turn would make it harder to just specify a default literal data type.

Oh well, can't have everything, I guess. But I do agree with Joan and Jadga, that the "empty" string value would be less bad than false.

**Comment by brianharvey** [comment]
I don't like using a value that's very likely to appear in a list as the signal that there is no satisfactory item in the list.  So I vote no on either kind of empty.  I want #Failed, an object that's not identical to anything else but counts as false in conditionals.  And I want the COND => feature, so you can say
![Untitled script pic](https://user-images.githubusercontent.com/3885074/82953397-8fb5fe80-9f5f-11ea-8cd6-533dedec4467.png)
![Untitled script pic (1)](https://user-images.githubusercontent.com/3885074/82953421-9a709380-9f5f-11ea-820d-a77e45e7021d.png)
This embraces the common user error (according to you) of using the result from FIND FIRST directly in an IF.  Some expert can fix that bug with =>.

**Comment by jguille2** [comment]
Uff... but one question, Brian...

![firstItem](https://user-images.githubusercontent.com/2926315/82958063-d4b85000-9fb4-11ea-87f9-f025fcd4ae8e.png)

should report the same as

![firstItem2](https://user-images.githubusercontent.com/2926315/82958090-e4379900-9fb4-11ea-8348-d3afb71abdfb.png)
?

Because the second one reports that "empty string".

Joan

**Comment by jmoenig** [comment]
yeah, and also:

![untitled script pic (24)](https://user-images.githubusercontent.com/2596285/82983822-22eb4480-9ff1-11ea-9e85-610402fe64d9.png)

We should be consistent. Reporting `false` isn't helping.

**Comment by ToonTalk** [comment]
Brian's program produces the error in the beta release with "Expecting Boolean but got a list"

One needs instead to do

![image](https://user-images.githubusercontent.com/2138696/83005292-345d3c80-a043-11ea-81a1-e13cb55b1a09.png)

Are we happy with that?

**Comment by jmoenig** [comment]
I think I'm happy with this. Mind you I'm the one who's wary about introducing type-checks to primitives for reasons of performance. But I see benefit in introducing them now for those blocks that also take lists as inputs.

**Comment by ToonTalk** [comment]
I am more worried about backwards compatibility.  It seems it is technically possible for old projects to keep old behaviours but it isn't clear what should happen if they are edited. And maybe more important whether users will get confused if they encounter different behaviours for blocks such as `if` or `not`. This is also a problem with some of the customisation settings. 

I'm not against change or progress but it does seem very hard to manage well.

**Comment by brianharvey** [comment]
If you want consistency, the way to get it is to have a distinguished #Failed object, not using any value that also has a legitimate meaning.  @jguille2: The reason FIND FIRST behaves differently from ITEM 1 OF KEEP is that ITEM behaves badly.  If out-of-bounds ITEM reported #Failed, then so would FIND FIRST.

NOT should always report #True or #False.  So NOT (NOT X)) is a way to turn truthy or falsy values into official Booleans.  NOT (NULL? X)) is a common old-days Lisp idiom used only to make programs self-documenting because the empty list was already false.

An advantage of #Failed is that we could have a user-settable option to throw an error when a primitive reports it (rather than put checks in each primitive).

**Comment by jmoenig** [comment]
Uh-oh. Brian, I disagree with almost everything here. Can we talk about this 1 on 1 first?

**Comment by ToonTalk** [comment]
Please post a summary of the discussion here. Coming from a Lisp background I tend to agree with Brian. 

This reminds me of a discussion I had with [Alan Perlis](https://en.wikipedia.org/wiki/Alan_Perlis) in the late 80s. He claimed that nearly all programming language design disagreements are due to the first programming languages of each discussant.

**Comment by rdococ** [comment]
If I may offer my own novice opinion on this, I think that any solution that uses a "failure value" (e.g. returning false, an empty string or a special "failed" value) will result in problems.

A project might want to perform any operation that might produce a failure value, and then store the result of that operation in another list, whether it succeeded or not. Then, when it tries to access that list, it cannot distinguish between the case where the operation failed, and the case where the item the project is trying to access actually does not exist.

I think that the most logically sound way to solve this problem is to add a "maybe" or "option" type. Unlike just having a special "failure" value, a "maybe" type would wrap around successful values too, distinguishing between the two cases. Unfortunately, it wouldn't be very intuitive for newer Snap!pers, and it would be just as clunky with the Snap! interface as supporting a "default value" input.
